24.09.26 유저 삭제, 수정 기능 추가

arcensia opened the pull request:
1. 테스트 코드 공통 변수로 변경
2. 유저 멤버 삭제, 수정기능 추가

arcensia committed:
1. JPA User repository 테스트 코트 추가

arcensia committed:
1. JPA User repository 테스트 코드 추가

arcensia committed:
1. JPA User 객체를 직접 생성하여 User객체에 대한 변경이 불가능하도록 수정함.

arcensia committed:
jpa 탈퇴 기능 추가

arcensia committed:
유저 삭제 기능에 필요한 TODO 추가

arcensia committed:
Merge branch 'develop' into feat/user

arcensia committed:
유저 모두 삭제, 수정기능 추가

arcensia committed:
test 공통 변수이름 통일

TrulyNotMalware commented on application/src/main/kotlin/app/discord/service/user/UserServiceImpl.kt:
```diff
@@ -28,7 +28,14 @@ class UserServiceImpl(
             return UserResult(status = UserResultStatus.SUCCESS, errorMessage = "")
         }
-        else TODO()
+        else {
+            val oldUser: User = userRepository.findUserWithNullException(userIdentifier = userRegisterEvent.userIdentifier)
+            oldUser.updateUserInfo(
+                leaveTime = userRegisterEvent.leaveTime
+            )
+            userRepository.updateUser(user = oldUser)
+            return UserResult(status = UserResultStatus.SUCCESS, errorMessage = "")
+        }
```
User Register (가입) 하는 로직에서 새 유저가 아닐 때 leaveTime만 업데이트 하는게 맞는건가요?
전체적으로 이해가 안되는 로직입니다.

TrulyNotMalware commented on domain/src/main/kotlin/app/discord/user/entity/User.kt:
```diff
@@ -63,6 +68,15 @@ class User (
         if (!validator(this)) throw IllegalArgumentException(errorMessage)
     }
 
+    private fun validateCheck(field: String, fieldName: String) {
+        field.validate(
+            { Pattern.matches(BASIC_NAME_REGEX, it) &&
+                    !Pattern.matches(ENGLISH_BAD_WORD_REGEX, it) &&
+                    !Pattern.matches(SQL_INJECTION_REGEX, it) },
+            errorMessage = "Invalid $fieldName"
+        )
+    }
+
```
fieldName 을 ErrorMessage 밖에 사용하는곳이 없는데, 그냥 Field 값을 넣어도 될것같아요.

TrulyNotMalware commented on infrastructure/src/main/kotlin/app/discord/listeners/BotRegisterGuildEventListener.kt:
```diff
@@ -0,0 +1,30 @@
+        nickname = member.nickname ?: "",
+        leaveTime = OffsetDateTime.MIN,
+    )
+}
```
Deprecated로 사용보류하려면 `@DiscordEventHandler` 어노테이션을 제거해주세요.

TrulyNotMalware commented on infrastructure/src/main/kotlin/app/discord/listeners/MessageListener.kt:
```diff
@@ -0,0 +1,20 @@
+        //TODO 메시지 받아서 명령어 수행하는 로직 추가할것. 1. 모든 유저를 등록하는 로직 추가.
+    }
+
+}
```
마찬가지로 어노테이션 제거해주세요.

TrulyNotMalware commented on infrastructure/src/main/kotlin/app/discord/repository/jpa/user/schema/UserEntity.kt:
```diff
@@ -2,40 +2,46 @@ package app.discord.repository.jpa.user.schema
 
     @field: Column(name = "USER_ID", nullable = false)
-    var userId: String,
+    val userId: String,
```
전체적으로 코드에서 UserId 와 GuildId 를 쌍으로 묶어서 조회하는 로직이 많이 보입니다.
이부분 `UserEntityIdentifier` 라는 [Embedded 객체가 있습니다.](https://github.com/StudyManagerBot/ManagerBot/blob/develop/infrastructure/src/main/kotlin/app/discord/repository/jpa/attendance/schema/UserEntityIdentifier.kt) 변경 고려해보세요.

```suggestion
    @field:Embedded
    val userIdentifier: UserEntityIdentifier,
```
[예제](https://github.com/StudyManagerBot/ManagerBot/blob/develop/infrastructure/src/main/kotlin/app/discord/repository/jpa/attendance/schema/JpaAttendanceHistoryEntity.kt)를 참고해주세요.

TrulyNotMalware commented on infrastructure/src/main/kotlin/app/discord/repository/jpa/user/schema/UserEntity.kt:
```diff
@@ -2,40 +2,46 @@ package app.discord.repository.jpa.user.schema
+    val leaveTime: OffsetDateTime
+
+)  {
+    override fun equals(other: Any?): Boolean {
+
+        return super.equals(other)
+    }
+
+    override fun hashCode(): Int {
+        return super.hashCode()
+    }
+}
```
재정의 안하려면 삭제해주세요!

TrulyNotMalware commented on infrastructure/src/main/kotlin/app/discord/repository/jpa/user/JpaUserEntityRepository.kt:
```diff
@@ -2,7 +2,10 @@ package app.discord.repository.jpa.user
+@Repository
 interface JpaUserEntityRepository: JpaRepository<UserEntity, Long>{
+    fun findAllByGuildId(guildId: String): List<UserEntity>
     fun findByGuildIdAndUserId(guildId:String, userId:String): UserEntity?
```
`UserIdentifier`사용을 고려해보세요.

TrulyNotMalware commented on infrastructure/src/main/kotlin/app/discord/repository/jpa/user/JpaUserRepository.kt:
```diff
@@ -66,7 +69,9 @@ class JpaUserRepository(
 
     override fun updateUser(user: User): User {
-        val userEntity = toJpaEntity(user)
+        val nowUser = jpaUserEntityRepository.findByGuildIdAndUserId(guildId = user.userIdentifier.guildId, userId = user.userIdentifier.userId)
```
마찬가지로 `UserIdentifier` 고려해보세요.

TrulyNotMalware commented on infrastructure/src/main/kotlin/app/discord/repository/jpa/user/JpaUserRepository.kt:
```diff
@@ -80,8 +85,15 @@ class JpaUserRepository(
     }
 
-    private fun toJpaEntity(user: User)=
+    override fun deleteUser(userIdentifier: UserIdentifier) {
+        val members = jpaUserEntityRepository.findAllByGuildId(guildId = userIdentifier.guildId)
+        jpaUserEntityRepository.deleteAllInBatch(members)
+    }
```
`UserIdentifier` 를 입력으로 받고 함수 이름은 deleteUser 입니다.
이 함수를 봤을 때 기대하는 기능은 UserIdentifier 로 조회한 단건 User 를 삭제하는 기능으로 인식됩니다.
하지만 내부 기능은 동일한 GuildId 를 가지는 모든 유저를 전체 삭제하는 실행합니다. 인식과 너무나도 다른 기능을 수행하므로 아주 위험해보입니다. **함수 이름을 좀더 직관적이고 명확하게 변경해주세요.**

또한 `deleteAll` 과 같은 함수는 매우 위험한 함수입니다. Marking 하는식으로 변경하는 것은 한번 고려해보세요.

TrulyNotMalware reviewed (changes requested):
repository 쪽 review 꼭 확인해주세요.

arcensia committed:
BotRegisterGuildEventListener, MessageListener
-> message 템플릿을 활용하므로 삭제함.
UserEntity: guildId, userId를 identifier로 embadded하여 사용
JpaUserEntityRepository: UserEntity변경점으로인한 코드수정
